src/wagtail_vector_index/storage/django.py

```diff
@@ -381,10 +468,36 @@ def bulk_generate_documents(self, objects, *, embedding_backend):
             ),
         )
 
+    # Interface methods
+    @transaction.atomic
     def to_documents(
         self, object: models.Model, *, embedding_backend: BaseEmbeddingBackend
     ) -> Generator[Document, None, None]:
-        yield from self.generate_documents(object, embedding_backend=embedding_backend)
+        """Use the AI backend to generate and store Documents for this object"""
+        chunks = self._get_chunks(object, embedding_backend)
+        documents = Document.objects.for_key(ModelKey(object))
+
+        if self._existing_documents_match(list(documents), chunks):
+            yield from documents
+
+        yield from self._create_new_documents(object, chunks, embedding_backend)
+
+    async def ato_documents(
```

Comment:
We're losing the transaction here - would it be possible to restructure so we generate the document instances async, then save them later? Just feels like we're risking some inconsistent state